### DISCUSSION TITLE
[cpp] Serialize ROS2 message for mcap outside of ROS2

### DISCUSSION THREAD

**Discussion by andrewwongagility** [foxglove/mcap#1116] [General] [answered]
Hello,
I'm trying to serialize a ROS2 message outside of ROS2 (using a custom middleware). I've provided some cpp toy code below that demonstrates what I'm trying to accomplish:

**Writing data to mcap file (shown below)**
1. Convert our custom image format to ROS2 raw image msg
2. Serialize the ROS2 image msg to a byte array (`std::array<uint8_t>`)

**Reading data from mcap file  (not shown below)**
1. I'd also like to deserialize this message to use within our custom middleware

```
#include std_msgs/ImageRaw.cpp // ROS2 Image Raw msg

ImageFrame* frame;  // our custom image format

std_msgs::ImageRaw ros2_img; // ROS2 image format

// Fill out ROS 2 message struct with out custom image format
ros2_img.width = frame->width;
ros2_img.height = frame->height;
...
ros2_img.data = frame->data;

// TODO: perform serialization
```

Does foxglove provide any cpp libraries that serialize a ROS2 message? So far, I've looked into the [serialization code in rclcpp](https://github.com/ros2/rclcpp/blob/rolling/rclcpp/src/rclcpp/serialization.cpp), included in ROS2, but it appears that rclcpp has lots of dependencies, making it cumbersome to include as a library in our software. I would prefer to use a library instead of writing the serialization/deserialization myself, because I think it is easier to upgrade a library than rewrite custom code in the event that the ROS2 message definitions change in the future.

Thank you for your help!

**Comment by jtbandes** [comment, marked as answer]
The only implementation we currently provide is in TypeScript, in [@foxglove/rosmsg2-serialization](https://github.com/foxglove/rosmsg2-serialization) which uses [@foxglove/cdr](https://github.com/foxglove/cdr) under the hood.

**Comment by andrewwongagility** [comment]
I apologize for the late reply. Thank you for your help; I've decided to use flatbuffers, since this is another workflow supported by Foxglove
